mlir/include/mlir/Dialect/MIOpen/MIOpenOpTraits.h

```diff
@@ -0,0 +1,31 @@
+//===- MIOpenOpTraits.h - MLIR MIOpen dialect operation traits --*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares C++ classes for some of operation traits in the MIOpen
+// dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_DIALECT_MIOPEN_OPTRAITS_H_
+#define MLIR_DIALECT_MIOPEN_OPTRAITS_H_
+
+#include "mlir/IR/OpDefinition.h"
+
+namespace mlir {
+namespace OpTrait {
+namespace miopen {
+
+/// An operation trait for convolution ops.
+template <typename ConcreteType>
+class ConvolutionOp : public TraitBase<ConcreteType, ConvolutionOp> {};
```

Comment:
First off, I'm not sure we need the `OpTrait` namespace here.

Second, ... can't we tablegen this whole thing?

Reply:
To my understanding, we need a subclass of `TraitBase` be defined in C++.

Reply:
Nope!

Here's how I'm adding an `InferIntRangeInterface` to `arith` in some background work, minus header guards

`InferIntRangeInterface.h`
```c++
#include "mlir/IR/Attributes.h"
#include "mlir/IR/OpDefinition.h"
#include "mlir/IR/Operation.h"

#include "mlir/Dialect/Arithmetic/IR/InferIntRangeInterface.h.inc"
```

`InferIntRangeInterface.td`

```
include "mlir/IR/OpBase.td"

def InferIntRangeInterface : OpInterface<"InferIntRangeInterface"> {
    let description = [{ ... }];
    let cppNamespace = "::mlir::arith";
    let methods = [...];
}
```
CMakeLists in the include directory
```
add_mlir_interface(InferIntRangeInterface)
```

`InferIntRangeInterface.cpp`
```c++
#include "mlir/Dialect/Arithmetic/IR/InferIntRangeInterface.h"
#include "mlir/Dialect/Arithmetic/IR/InferIntRangeInterface.cpp.inc"
```

And that's it!